### ISSUE TITLE
High memory usage: from atproto import Client

### ISSUE THREAD

**Issue by mtxvp** [MarshalX/atproto#316]
Merely adding the following line to an empty python script makes memory usage to jump from 0 to 100+MiB:

```
from atproto import Client
```

Tested using the following command on WSL2, python 3.10.6:

```
mprof run python -m experiments.profilecode   &&  mprof plot -o image.png
```
(see [memory profiler](https://pypi.org/project/memory-profiler/), section: Time-based memory usage)

![image](https://github.com/MarshalX/atproto/assets/150715589/d8aaca01-fcff-4164-ab32-df05544bb521)

**Comment by MarshalX** [comment]
Thank you for your report. Here is why it happens: https://github.com/pydantic/pydantic/issues/6620#issuecomment-1934687870

TLDR: pydantic copies some stuff into the rust side and it consumes memory; upgrade to the latest pydantic

So nothing I can do from my side IMO. You can try to load only the necessary models for your project if the memory usage is critical.

**Comment by mtxvp** [comment]
> Thank you for your report. Here is why it happens: [pydantic/pydantic#6620 (comment)](https://github.com/pydantic/pydantic/issues/6620#issuecomment-1934687870)
> 
> TLDR: pydantic copies some stuff into the rust side and it consumes memory; upgrade to the latest pydantic
> 
> So nothing I can do from my side IMO. You can try to load only the necessary models for your project if the memory usage is critical.

Validation or having access to models is not critical at all or necessary in my use case. The less memory I can use, the better.

Given my limited understanding on how to use your library, I'd love to get some recommendations on how load only minimal amount of models, if any. All I am trying to do is to use API for posting by following your README, which suggests importing Client.

Importing Client in turn executes the following line of code, which seems to be the on of the culprits:

atproto_client/models/__init__.py
```
...
...
load_models()
```

Also, latest pydantic v2.6.4 is added to my project via dependencies, I have no direct use of it otherwise.

**Comment by MarshalX** [comment]
The current situation is even worse. I tried your commands and it gives me ~170mb on `from atproto_client import models`.

Even when model rebuilding (in `load_models`) is disabled (which breaks the code because not all models are in completed state) it consumes around 50mb just by importing all models in the `model/__init__.py.

A little about `load_models`. Because of so deep, nested, and sometimes a bit recursive model structure this is important to load models in a strict order to resolve dependencies of each other. Keeping such order could be hard to achieve from our code generator. Which is a fully automated process. That's why model rebuilding happens in runtime when all the models are defined already. `load_models` at the end calls `model_rebuild` for each existing model but the `model_rebuild` method is pretty smart to does nothing if the model is completely loaded (all types are resolved correctly). So there is no overhead. 

TL;DR: I think there is no way to fix it from our side except migrating from pydantic to something else. `load_models` is not a problem. The problem is **470 models** which are inherited from `pydantic.BaseModel` with Rust\`s memory overhead for **each** of them. IMO only pydantic\`s team can try to minimize their memory overhead

**Comment by MarshalX** [comment]
I've created a separate isolated repository that contains pure models without additional noise and asked pydantic team about the amount of consumed memory here: https://github.com/pydantic/pydantic/issues/9982

**Comment by toughyear** [comment]
is there any solution here @MarshalX ?/

**Comment by MarshalX** [comment]
@toughyear I do not think that this is possible to fix without effort from pydantic team :(

**Comment by toughyear** [comment]
appreciate the quick response @MarshalX !

**Comment by MarshalX** [comment]
pydantic team did a very good job in the v2.11 release. the importing time and memory usage have been improved! 

- their release with details: https://pydantic.dev/articles/pydantic-v2-11-release
- my comparison with models of atproto and bluesky: https://github.com/MarshalX/pydantic-memory-usage/commit/f3f2d3600ba1148d4abbf8961a128f028c8fdb9a

i highly recommend updating pydantic; gonna close the issue

**Comment by toughyear** [comment]
@MarshalX thanks for updating on this!

So memory consumption seems halved with the new release. Would you suggest trying again? any other improvements we can do for low memory usage?

**Comment by jvanasco** [comment]
Could performance be improved by caching the pydantic models to/from json?  The models should only need to be built once per atproto release, only project developer should need to worry about cache expiry.

**Comment by MarshalX** [comment]
@jvanasco i do not think that this is possible. Pydantic "builds" models into internal structures. Which are stored just in runtime. Even inside Rust structures, not Python ones (pydantic-core). This is exactly what uses memory. And be able to dump/load this memory to speed up import time is something difficult because they should care then about compatibility of internal structs between pydantic versions etc and somehow validate outdated cache. IMHO

**Comment by jvanasco** [comment]
Thanks for the reply.  I’m not very familiar with Pydantic, and was likely confused by some of the articles and blogs on optimizing it after looking into this.There were some optimization articles that leveraged pydantic’s json serialization of models.If that were possible (looking at things now, more likely not applicable), in terms of updating across versions and rebuilding caches, that’s not an issue.  A lot of libraries manage precomputed caches via version based directories in the user space Python directory.   Offhand, I know tldextract caches raw and processed versions of the Public Suffix List on disk and supports rebuilding; I am pretty sure it is in a versioned directory, but I should probably check and submit a PR if not.  So I think your concerns could easily be addressed if this were technically possible (but it doesn’t seem to be).  On Nov 9, 2025, at 4:54 AM, Ilya Siamionau ***@***.***> wrote:﻿MarshalX left a comment (MarshalX/atproto#316)
@jvanasco i do not think that this is possible. Pydantic "builds" models into internal structures. Which are stored just in runtime. Even inside Rust structures, not Python ones (pydantic-core). This is exactly what uses memory. And be able to dump/load this memory to speed up import time is something difficult because they should care then about compatibility of internal structs between pydantic versions etc and somehow validate outdated cache. IMHO
